File: Python/278.First Bad Version.py
#找到一堆版本中第一个出错的版本,(第一个出错后边都错)
#二分法
class Solution(object):
    def firstBadVersion(self, n):
        l , r = 0 , n
        while l<=r:
            mid = (l+r)//2
            if self.isBadVersion(mid):  r = mid-1
            else:   l = mid+1
        return l

    def isBadVersion(self,version):
        return version>3

s = Solution()
print(s.firstBadVersion(6))

File: C++/Binary Search/278.First Bad Version.cpp
/*
调用API函数isBadVersion寻找所有版本中第一次出现错误的版本。
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int firstBadVersion(int n) {
        int left = 1, right = n;
        while(left < right)
        {
        	int mid = left + (right-left)/2;
        	if(isBadVersion(mid))	right = mid;
        	else left = mid + 1;
        }
        return left;
    }
};

int main()
{
	Solution s;
	cout << s.firstBadVersion(10) << endl;
}

